src/pj_inv.c

```diff
@@ -1,60 +1,167 @@
+ *
+ ******************************************************************************
+ * Copyright (c) 2000, Frank Warmerdam
+ * Copyright (c) 2018, Thomas Knudsen / SDFE
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *****************************************************************************/
 #include <errno.h>
-# define EPS 1.0e-12
 
-/* inverse projection entry */
-LP pj_inv(XY xy, PJ *P) {
-    LP lp;
-    LP err;
-    int last_errno;
+#include "proj_internal.h"
+#include "projects.h"
 
-    /* cannot const-initialize this due to MSVC's broken (non const) HUGE_VAL */
-    err.lam = err.phi = HUGE_VAL;
 
-    if (0==P->inv)
-        return err;
 
-    /* can't do as much preliminary checking as with forward */
-    if (xy.x == HUGE_VAL || xy.y == HUGE_VAL) {
-        pj_ctx_set_errno( P->ctx, -15);
-        return err;
+PJ_COORD pj_inv_prepare (PJ *P, PJ_COORD coo) {
+    if (coo.xyz.x == HUGE_VAL) {
+        proj_errno_set (P, PJD_ERR_INVALID_X_OR_Y);
+        return proj_coord_error ();
     }
 
-    last_errno = proj_errno_reset (P);
-
     /* de-scale and de-offset */
-    xy.x = (xy.x * P->to_meter - P->x0);
-    xy.y = (xy.y * P->to_meter - P->y0);
+    coo.xyz.x = (coo.xyz.x * P->to_meter  - P->x0);
+    coo.xyz.y = (coo.xyz.y * P->to_meter  - P->y0);
+    coo.xyz.z = (coo.xyz.z * P->vto_meter - P->z0);
 
-    /* Classic proj.4 functions expect plane coordinates in units of the semimajor axis */
-    /* Multiplying by ra, rather than dividing by a because the CALCOFI projection      */
-    /* stomps on a and hence depends on this */
+    /* Classic proj.4 functions expect plane coordinates in units of the semimajor axis  */
+    /* Multiplying by ra, rather than dividing by a because the CALCOFI projection       */
+    /* stomps on a and hence (apparently) depends on this to roundtrip correctly         */
+    /* (CALCOFI avoids further scaling by stomping - a better solution must be possible) */
```

Comment:
It (probably) is, since we have better ways of avoiding unexpected scaling now (not setting P->right to classic), but I will have to check with the original author of the CALCOFI code, as they stomp on more than just `a`

Reply:
(note that the CALCOFI stuff is not related to this PR - it has been in for a lo-ong time)